### ISSUE TITLE
Should we make this header-only? 

### ISSUE THREAD

**Issue by bmagyar** [ros-controls/urdf_geometry_parser#10]
I think we could simplify a couple of controllers but I'd opt for not linking to a lib for it.
Opinions @ros-controls/ros_control ?

**Comment by bmagyar** [comment]
@artivis this could shorten the `diff_drive_controller` code too

**Comment by matthew-reynolds** [comment]
Reviving this to point something small out:

Beyond the normal considerations of header-only (longer compiles, but more chance for the compiler to optimize, etc), `tf2` is a build-only dependency (At least since #14), so if this were header-only that would add a little overhead. Realistically though, I would be surprised if there are many cases where you'd want `urdf_geometry_parser` without `tf2`.

**Comment by bmagyar** [comment]
After a quick glance at the code... it's not using roscpp anymore and even tf2 is a very very marginal, case, used internally so probably dependencies could be reduced to urdf. Much larger of a chance for a header-only now! ;)

**Comment by matthew-reynolds** [comment]
We still need `roscpp` for our rosconsole outputs (`ROS_ERROR`, `ROS_DEBUG`, etc), and for fetching the robot description from the param server. But you're right, we definitely don't need `tf2`: See #15.

**Comment by bmagyar** [comment]
You are right... While the purist angel on one shoulder says "we could turn this into a C++ header-only library grabbing a string and returning error strings when failing" the little devil on the other one prefers convenience libraries to be...convenient. I side with the devil today

**Comment by artivis** [comment]
Note that fetching the param from the param server (`nh.getParam`) could (should?) be move outside the constructor. After all, it is an `UrdfGeometryParser` class not an `UrdfGeometryFetcherAndParser`. 

As for the logging, one could replace it with an output argument e.g. `UrdfGeometryParser(const std::string& robot_model_str, const std::string& base_link, UrdfGeometryParserException& exception)`. Or simply throw `UrdfGeometryParserException` (with `class UrdfGeometryParserException : std::runtime_error`).

I guess implementing the above two propositions implies important changes, `Noetic`?

@bmagyar looks like I'm playing on the angel of template side ;)
